api/src/OpenTelemetry/Logs/Core.hs

```diff
@@ -124,41 +126,56 @@ setGlobalLoggerProvider :: (MonadIO m) => LoggerProvider body -> m ()
 setGlobalLoggerProvider = liftIO . writeIORef globalLoggerProvider
 
 
+defaultShutdownTimeout :: Int
+defaultShutdownTimeout = 5_000_000
+
+
 {- | This method provides a way for provider to do any cleanup required.
 
  This will also trigger shutdowns on all internal processors.
 -}
-shutdownLoggerProvider :: (MonadIO m) => LoggerProvider body -> m ()
-shutdownLoggerProvider LoggerProvider {loggerProviderProcessors} = liftIO $ do
-  asyncShutdownResults <- V.forM loggerProviderProcessors $ \processor -> do
-    logRecordProcessorShutdown processor
-  mapM_ wait asyncShutdownResults
+shutdownLoggerProvider
+  :: (MonadIO m)
+  => Maybe Int
+  -- ^ Optional timeout in microseconds, defaults to 5,000,000 (5s)
+  -> LoggerProvider body
+  -> m ShutdownResult
+  -- ^ Result that denotes whether the shutdown action succeeded, failed, or timed out.
+shutdownLoggerProvider mtimeout LoggerProvider {loggerProviderProcessors} = liftIO $ do
+  mresult <-
+    timeout (fromMaybe defaultShutdownTimeout mtimeout) $
+      takeWorstShutdownResult
+        <$> forConcurrently
+          loggerProviderProcessors
+          (\lrp -> logRecordProcessorShutdown lrp `catch` \(SomeException _) -> pure ShutdownFailure)
+
+  case mresult of
+    Nothing -> pure ShutdownTimeout
+    Just res -> pure res
+
+
+defaultForceFlushTimeout :: Int
+defaultForceFlushTimeout = 5_000_000
 
 
 {- | This method provides a way for provider to immediately export all @LogRecord@s that have not yet
  been exported for all the internal processors.
 -}
 forceFlushLoggerProvider
   :: (MonadIO m)
-  => LoggerProvider body
-  -> Maybe Int
+  => Maybe Int
   -- ^ Optional timeout in microseconds, defaults to 5,000,000 (5s)
+  -> LoggerProvider body
   -> m FlushResult
   -- ^ Result that denotes whether the flush action succeeded, failed, or timed out.
-forceFlushLoggerProvider LoggerProvider {loggerProviderProcessors} mtimeout = liftIO $ do
-  jobs <- V.forM loggerProviderProcessors $ \processor -> async $ do
-    logRecordProcessorForceFlush processor
+forceFlushLoggerProvider mtimeout LoggerProvider {loggerProviderProcessors} = liftIO $ do
   mresult <-
-    timeout (fromMaybe 5_000_000 mtimeout) $
-      V.foldM
-        ( \status action -> do
-            res <- waitCatch action
-            pure $! case res of
-              Left _err -> FlushError
-              Right _ok -> status
-        )
-        FlushSuccess
-        jobs
+    timeout (fromMaybe defaultForceFlushTimeout mtimeout) $
+      takeWorstFlushResult
+        <$> forConcurrently
+          loggerProviderProcessors
+          (\lrp -> logRecordProcessorForceFlush lrp `catch` \(SomeException _) -> pure FlushError)
```

Comment:
The combinators `takeWorstFlushResult` and `forConcurrently` make this code a bit cleaner than before.